File: src/commands/amend.ts
import { existsSync, readFileSync } from "node:fs";
import { runPaths } from "../core/paths.js";
import { diffNoIndex } from "../core/git.js";
import { hashPlanFile, parsePlanFile } from "../artifacts/plan.js";
import { nowIso, writeRun } from "../core/run.js";
import { emit, GateError, requireActiveRun, type ParsedArgs } from "./shared.js";

/**
 * `gate amend` - show the diff between the current plan.md and the last
 * approved snapshot, and record intent to re-approve it (Milestone 5). Does
 * not itself re-approve - that's `gate approve --amend` - so a delta
 * re-approval always happens after a human has actually seen the diff, the
 * same discipline `gate approve` already applies to the initial sign-off
 * (writing the plan and approving it are two separate, deliberate steps).
 */
export function cmdAmend(args: ParsedArgs): void {
  const { root, run } = requireActiveRun(args);
  if (!run.approval) {
    throw new GateError("plan was never approved - nothing to amend; run `gate approve` first");
  }

  const { plan: planPath, planApproved: approvedPath } = runPaths(root, run.id);
  const { plan, errors } = parsePlanFile(planPath);
  if (!plan) {
    throw new GateError(`cannot amend an invalid plan: ${errors.join("; ")}`);
  }
  const currentHash = hashPlanFile(planPath);
  if (!currentHash) throw new GateError("plan.md not found");
  if (currentHash === run.approval.planHash) {
    throw new GateError("plan.md matches the approved version - nothing to amend");
  }

  const { diff, trustedSnapshot, warning } = renderDiff(approvedPath, planPath, run.approval.planHash);

  const by =
    (typeof args.flags.by === "string" ? args.flags.by : undefined) ??
    process.env.GATE_SESSION_ID ??
    null;

  run.amendment = { planHash: currentHash, at: nowIso(), by };
  writeRun(root, run);

  const human = [
    `Amendment recorded${by ? ` by ${by}` : ""}.`,
    warning ?? "Diff vs the approved plan:",
    "",
    diff.trim() || "(no textual diff produced)",
    "",
    "Run `gate approve --amend` to re-approve this delta.",
  ].join("\n");
  emit(human, { amended: true, diff, trustedSnapshot, warning, ...run.amendment }, args.flags);
}

/**
 * The diff `gate amend` shows, only ever computed from a snapshot proven to
 * be the plan `gate approve` actually recorded (review finding F2): trusting
 * `plan.approved.md` on its face would let a doctored snapshot produce an
 * empty or misleading diff - e.g. a snapshot rewritten to match the
 * *current* plan.md would show "no changes" for a plan that in fact drifted
 * far from what was approved, and a human re-approving on that diff approves
 * blind. The snapshot is trustworthy only when its own hash matches
 * `approval.planHash` - the same value `gate approve` bound at sign-off, and
 * the same one the drift check (`gates/planDrift.ts`) already compares
 * against. Anything else (missing, unreadable, or hash-mismatched) falls
 * back to showing the full current plan with a loud warning, never a diff
 * computed against content that can't be trusted.
 */
function renderDiff(
  approvedPath: string,
  planPath: string,
  approvedHash: string,
): { diff: string; trustedSnapshot: boolean; warning: string | null } {
  if (!existsSync(approvedPath)) {
    return {
      diff: `(no approved snapshot on disk - showing the full current plan)\n\n${readFileSync(planPath, "utf8")}`,
      trustedSnapshot: false,
      warning: null,
    };
  }
  const snapshotHash = hashPlanFile(approvedPath);
  if (snapshotHash !== approvedHash) {
    return {
      diff: readFileSync(planPath, "utf8"),
      trustedSnapshot: false,
      warning:
        "WARNING: the approved-plan snapshot on disk does not match the recorded approval hash " +
        "(edited or corrupted since `gate approve` wrote it) - showing the full current plan " +
        "instead of an untrustworthy diff. Review it in full before re-approving.",
    };
  }
  return { diff: diffNoIndex(approvedPath, planPath), trustedSnapshot: true, warning: null };
}

File: rust/src/commands/amend.rs
//! Port of `src/commands/amend.ts`.
//!
//! `gate amend` - show the diff between the current plan.md and the last
//! approved snapshot, and record intent to re-approve it (Milestone 5).
//! Does not itself re-approve - that's `gate approve --amend` - so a delta
//! re-approval always happens after a human has actually seen the diff,
//! the same discipline `gate approve` already applies to the initial
//! sign-off.

use std::fs;
use std::path::Path;

use crate::artifacts::plan::{hash_plan_file, parse_plan_file};
use crate::cli::args::parse_args;
use crate::cli::context::require_active_run;
use crate::cli::output::{emit, UserError};
use crate::core::git::diff_no_index;
use crate::core::json::Value;
use crate::core::paths::run_paths;
use crate::core::run::{now_iso, write_run, Amendment};

struct DiffResult {
    diff: String,
    trusted_snapshot: bool,
    warning: Option<String>,
}

/// The diff `gate amend` shows, only ever computed from a snapshot proven
/// to be the plan `gate approve` actually recorded (review finding F2):
/// trusting `plan.approved.md` on its face would let a doctored snapshot
/// produce an empty or misleading diff. The snapshot is trustworthy only
/// when its own hash matches `approval.planHash`; anything else (missing,
/// unreadable, or hash-mismatched) falls back to showing the full current
/// plan with a loud warning, never a diff computed against content that
/// can't be trusted.
fn render_diff(approved_path: &Path, plan_path: &Path, approved_hash: &str) -> DiffResult {
    if !approved_path.exists() {
        let current = fs::read_to_string(plan_path).unwrap_or_default();
        return DiffResult {
            diff: format!(
                "(no approved snapshot on disk - showing the full current plan)\n\n{current}"
            ),
            trusted_snapshot: false,
            warning: None,
        };
    }
    let snapshot_hash = hash_plan_file(approved_path);
    if snapshot_hash.as_deref() != Some(approved_hash) {
        let current = fs::read_to_string(plan_path).unwrap_or_default();
        return DiffResult {
            diff: current,
            trusted_snapshot: false,
            warning: Some(
                "WARNING: the approved-plan snapshot on disk does not match the recorded approval hash \
                 (edited or corrupted since `gate approve` wrote it) - showing the full current plan \
                 instead of an untrustworthy diff. Review it in full before re-approving."
                    .to_string(),
            ),
        };
    }
    DiffResult {
        diff: diff_no_index(
            &approved_path.to_string_lossy(),
            &plan_path.to_string_lossy(),
        ),
        trusted_snapshot: true,
        warning: None,
    }
}

pub fn run(argv: Vec<String>) -> Result<(), UserError> {
    // See `check::run`'s comment: prepend a placeholder command token so
    // `parse_args` never misreads a leading flag/positional as the command.
    let mut full = vec!["amend".to_string()];
    full.extend(argv);
    let args = parse_args(&full);
    let ctx = require_active_run(Some(&args))?;
    let mut run = ctx.run;
    let Some(approval) = run.approval.clone() else {
        return Err(UserError::new(
            "plan was never approved - nothing to amend; run `gate approve` first",
        ));
    };

    let paths = run_paths(&ctx.root, &run.id);
    let parsed = parse_plan_file(&paths.plan);
    if parsed.plan.is_none() {
        return Err(UserError::new(format!(
            "cannot amend an invalid plan: {}",
            parsed.errors.join("; ")
        )));
    }
    let Some(current_hash) = hash_plan_file(&paths.plan) else {
        return Err(UserError::new("plan.md not found"));
    };
    if current_hash == approval.plan_hash {
        return Err(UserError::new(
            "plan.md matches the approved version - nothing to amend",
        ));
    }

    let DiffResult {
        diff,
        trusted_snapshot,
        warning,
    } = render_diff(&paths.plan_approved, &paths.plan, &approval.plan_hash);

    let by = args
        .flags
        .str("by")
        .map(String::from)
        .or_else(|| std::env::var("GATE_SESSION_ID").ok());

    run.amendment = Some(Amendment {
        plan_hash: current_hash,
        at: now_iso(),
        by: by.clone(),
    });
    write_run(&ctx.root, &mut run).map_err(|e| UserError::new(e.to_string()))?;

    let by_suffix = by
        .as_deref()
        .filter(|s| !s.is_empty())
        .map(|s| format!(" by {s}"))
        .unwrap_or_default();
    let diff_trimmed = diff.trim();
    let human = [
        format!("Amendment recorded{by_suffix}."),
        warning
            .clone()
            .unwrap_or_else(|| "Diff vs the approved plan:".to_string()),
        String::new(),
        if diff_trimmed.is_empty() {
            "(no textual diff produced)".to_string()
        } else {
            diff_trimmed.to_string()
        },
        String::new(),
        "Run `gate approve --amend` to re-approve this delta.".to_string(),
    ]
    .join("\n");

    let amendment = run.amendment.as_ref().unwrap();
    let mut data = Value::object();
    data.insert("amended", true);
    data.insert("diff", diff.as_str());
    data.insert("trustedSnapshot", trusted_snapshot);
    data.insert("warning", warning.clone());
    data.insert("planHash", amendment.plan_hash.as_str());
    data.insert("at", amendment.at.as_str());
    data.insert("by", amendment.by.clone());

    emit(&human, &data, &args.flags)
}
